File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/OriginDeserializer.java
package io.appium.espressoserver.lib.helpers.w3c.models;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.Action.ELEMENT_CODE;
import static io.appium.espressoserver.lib.helpers.w3c.models.InputSource.VIEWPORT;

public class OriginDeserializer implements JsonDeserializer<Origin> {

    @Override
    public Origin deserialize(JsonElement json, Type paramType,
                                   JsonDeserializationContext paramJsonDeserializationContext) throws JsonParseException {
        Origin origin = new Origin();

        if (json.isJsonPrimitive()) {
            origin.setType(json.getAsString());
        } else if(json.isJsonObject() &&
                json.getAsJsonObject().get(ELEMENT_CODE).isJsonPrimitive()) {
            String elementId = json.getAsJsonObject().get(ELEMENT_CODE).getAsString();
            origin.setElementId(elementId);
            origin.setType(ELEMENT_CODE);
        } else if (json.isJsonNull()) {
            origin.setType(VIEWPORT);
        }

        return origin;
    }
}

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/OriginDeserializer.kt
package io.appium.espressoserver.lib.helpers.w3c.models

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type

class OriginDeserializer : JsonDeserializer<Origin> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, paramType: Type,
                             paramJsonDeserializationContext: JsonDeserializationContext): Origin {
        val origin = Origin()
        if (json.isJsonPrimitive) {
            origin.type = json.asString
        } else if (json.isJsonObject &&
                json.asJsonObject[InputSource.Action.ELEMENT_CODE].isJsonPrimitive) {
            val elementId = json.asJsonObject[InputSource.Action.ELEMENT_CODE].asString
            origin.elementId = elementId
            origin.type = InputSource.Action.ELEMENT_CODE
        } else if (json.isJsonNull) {
            origin.type = VIEWPORT
        }
        return origin
    }
}